src/NUnitFramework/framework/Constraints/ExactCountConstraint.cs

```diff
@@ -31,9 +31,19 @@ namespace NUnit.Framework.Constraints
     /// item in a collection, succeeding only if a specified
     /// number of items succeed.
     /// </summary>
-    public class ExactCountConstraint : PrefixConstraint
+    public class ExactCountConstraint : Constraint
```

Comment:
Well, originally the property constraint was the same way: it either did one thing or two things, depending on whether a constraint followed it. It was refactored into two constraints. One of them does one thing (checks that the property exists) while the other does two (checks that it exists and applies a constraint to it).

The disadvantage of the approach is that the existence check has to be in two different places. The advantage is a somewhat cleaner code line in the constraint. The test for whether there is a constraint would have to be done in at least two places in the code of a single class: once for applying the constraint and again for the message.

The big advantage though is a potential one in the design of operators. Reading your code, I got to thinking that we could generalize operators like prop and exactly. They would produce one of two constraints depending on context. A derived operator would only have to specify the Types of the two constraints. This is just an idea fluttering in my head, however. I haven't worked it out in code. But it opens the door to user-created operators as well as constraints, so that's interesting.